### PR TITLE
[cxx-interop] Implement foreign reference types.

### PR DESCRIPTION
**Pull request by zoecarver**
This is an experimental feature to allow an attribute, `import_as_ref`, to import a C++ record as a non-reference-counted reference type in Swift.

### CODE REVIEW

File: lib/ClangImporter/ImportDecl.cpp [hunk 1]
```diff
@@ -4276,9 +4295,10 @@ namespace {
         result = func;
 
         if (!dc->isModuleScopeContext()) {
-          if (selfIsInOut)
+          // Don't import members of a class decl as mutating.
+          if (selfIsInOut && !isa<ClassDecl>(dc)) {
```
Comment (rjmccall) [inline lib/ClangImporter/ImportDecl.cpp:4299]: Maybe `selfIsInOut` should just be false when we import as a reference type.

File: lib/IRGen/GenCall.cpp [hunk 2]
```diff
@@ -3510,6 +3529,13 @@ static void externalizeArguments(IRGenFunction &IGF, const Callee &callee,
     case clang::CodeGen::ABIArgInfo::IndirectAliased:
       llvm_unreachable("not implemented");
     case clang::CodeGen::ABIArgInfo::Indirect: {
+      // Classes are already indirect so we don't need to do anything here.
+      if (paramType.isForeignReferenceType()) {
+        assert(!silConv.isSILIndirect(params[i - firstParam]));
+        out.add(in.claimNext());
+        break;
+      }
+
```
Comment (rjmccall) [inline lib/IRGen/GenCall.cpp:3538]: We shouldn't see this.

**Review by rjmccall — CHANGES_REQUESTED** [review]

**Comment by zoecarver** [comment]
@rjmccall this should be ready for re-review now. I think I addressed all your comments. Let me know what you think.

Note: I still need to fix the issue with optionals (unless some macro tells us otherwise, we should be importing foreign reference types as optionals). I'll do that shortly, but I think the rest of the PR can still be reviewed without that change.

File: lib/SIL/Verifier/SILVerifier.cpp [hunk 3]
```diff
@@ -3351,9 +3353,14 @@ class SILVerifier : public SILVerifierBase<SILVerifier> {
...
     require(!isa<ExtensionDecl>(member.getDecl()->getDeclContext()),
             "extension method cannot be dispatched natively");
-    
-    // The method ought to appear in the class vtable.
-    require(VerifyClassMethodVisitor(member).Seen,
+
+    // The method ought to appear in the class vtable unless it's a foreign
+    // reference type.
+    bool isForeignReferenceTy =
+        operandType.getClassOrBoundGenericClass()->getClangDecl() &&
```
[the remark this answers is not in the document]
Comment (zoecarver) [reply lib/SIL/Verifier/SILVerifier.cpp:3360]: But thanks for catching that!

**Comment by zoecarver** [comment]
@swift-ci please smoke test.

File: include/swift/AST/Decl.h [hunk 4]
```diff
@@ -3841,7 +3842,8 @@ class ClassDecl final : public NominalTypeDecl {
   ///
   /// \see getForeignClassKind
   bool isForeign() const {
-    return getForeignClassKind() != ForeignKind::Normal;
+    return getForeignClassKind() != ForeignKind::Normal ||
+      const_cast<ClassDecl *>(this)->isForeignReferenceType();
```
Comment (zoecarver) [inline include/swift/AST/Decl.h:3846]: This is not ideal, but I think it's OK because we never construct a const decl, right?

**Comment by zoecarver** [comment]
@swift-ci please smoke test.

**Comment by zoecarver** [comment]
@swift-ci please smoke test.

**Comment by zoecarver** [comment]
@rjmccall friendly ping :)

File: include/swift/AST/DiagnosticsIRGen.def [hunk 5]
```diff
@@ -59,5 +59,10 @@ ERROR(temporary_allocation_alignment_not_positive,none,
 ERROR(temporary_allocation_alignment_not_power_of_2,none,
       "alignment value must be a power of two", ())
 
+ERROR(foreign_reference_types_unsupported,none,
```
Comment (zoecarver) [inline include/swift/AST/DiagnosticsIRGen.def:62]: @rjmccall and @beccadax I need to add a test for this new error, but other than that, this should be ready for re-review.

**Comment by zoecarver** [comment]
@swift-ci please smoke test.

**Comment by zoecarver** [comment]
@swift-ci please smoke test.

**Comment by zoecarver** [comment]
@swift-ci please smoke test.

**Comment by zoecarver** [comment]
@swift-ci please smoke test.

**Comment by zoecarver** [comment]
@swift-ci please smoke test Linux.

**Comment by zoecarver** [comment]
@swift-ci please smoke test Linux.

**Comment by zoecarver** [comment]
@swift-ci please smoke test Linux.

**Comment by zoecarver** [comment]
@swift-ci please smoke test Linux.

**Comment by zoecarver** [comment]
@swift-ci please smoke test.

**Comment by zoecarver** [comment]
@swift-ci please smoke test.

File: lib/AST/Type.cpp [hunk 6]
```diff
@@ -5612,6 +5608,18 @@ TypeBase::getAutoDiffTangentSpace(LookupConformanceFn lookupConformance) {
...
 }
 
+bool TypeBase::isForeignReferenceType() {
+  if (auto *classDecl = lookThroughAllOptionalTypes()->getClassOrBoundGenericClass())
+    return classDecl->isForeignReferenceType();
+  return false;
+}
+
+bool CanType::isForeignReferenceType() {
+  if (auto *classDecl = getPointer()->lookThroughAllOptionalTypes()->getClassOrBoundGenericClass())
```
Comment (hyp) [inline lib/AST/Type.cpp:5618]: Looks like you duplicate the logic here. Can you call `getPointer()->isForeignReferenceType()` instead?

File: test/Interop/Cxx/foreign-reference/not-any-object.swift [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+};
+
+//--- test.swift
+
+import Test;
+
+public func test(_ _: AnyObject) {}
+
+// TODO: make this a better error.
+test(Empty.create()) // expected-error {{type of expression is ambiguous without more context}}
```
Comment (zoecarver) [inline test/Interop/Cxx/foreign-reference/not-any-object.swift:29]: Add a test for casting explicitly to any object.

**Review by hyp — APPROVED** [review]

**Pull request merged**